flocker/testtools/test/test_flaky.py

```diff
@@ -42,63 +133,167 @@ def test_successful_flaky_test(self):
-        'unexpectedSuccesses': len(result.unexpectedSuccesses),
-        'testsRun': result.testsRun,
-    }
+        class SomeTest(AsyncTestCase):
 
+            run_tests_with = silent_async_runner
 
-def get_results(test):
-    """
-    Run a test and return a summary of its results.
-    """
-    result = unittest.TestResult()
-    test.run(result)
-    return _get_result_stats(result)
+            @flaky(u'FLOC-XXXX', max_runs=len(test_methods), min_passes=1)
+            def test_something(self):
+                next(executions)()
+
+        test = SomeTest('test_something')
+        self.assertThat(run_test(test), has_results(tests_run=Equals(1)))
+
+    @given(permutations([
+        lambda: throw(ValueError('failure')),
+        lambda: None,
+        lambda: throw(RuntimeError('failure #2')),
+    ]))
+    def test_intermittent_flaky_test_that_errors(self, test_methods):
+        """
+        Tests marked with 'flaky' are retried if they fail, and marked as
+        erroring / failing if they don't reach the minimum number of successes.
+        """
+        executions = iter(test_methods)
+
+        class SomeTest(testtools.TestCase):
+            run_tests_with = retry_flaky(output=StringIO())
+
+            @flaky(u'FLOC-XXXX', max_runs=len(test_methods), min_passes=2)
+            def test_something(self):
+                next(executions)()
+
+        test = SomeTest('test_something')
+        self.assertThat(run_test(test), has_results(
+            tests_run=Equals(1),
+            errors=HasLength(1),
+        ))
+
+    @given(permutations([
+        lambda: throw(ValueError('failure')),
+        lambda: None,
+        lambda: throw(RuntimeError('failure #2')),
+    ]))
+    def test_intermittent_flaky_subclass(self, test_methods):
+        """
+        We sometimes subclass test classes in order to test different
+        implementations of the same interface. A test within such a subclass
+        can be marked as flaky, causing it to retry.
+        """
+        executions = iter(test_methods)
+
+        class SomeTest(testtools.TestCase):
+            run_tests_with = retry_flaky(output=StringIO())
+
+            def test_something(self):
+                next(executions)()
+
+        class SubclassTest(SomeTest):
+
+            @flaky(u'FLOC-XXXX', max_runs=len(test_methods), min_passes=1)
+            def test_something(self):
+                super(SubclassTest, self).test_something()
+
+        test = SubclassTest('test_something')
+        self.assertThat(run_test(test), has_results(tests_run=Equals(1)))
+
+    @given(jira_keys, num_runs, num_runs,
+           streaming(text(average_size=10)).map(iter))
+    def test_flaky_skipped_test(self, jira_keys, max_runs, min_passes,
+                                reasons):
+        """
+        If a test is skipped and also marked @flaky, we report it as skipped.
+        """
+        [min_passes, max_runs] = sorted([min_passes, max_runs])
```

Comment:
More-or-less recommended practice: http://hypothesis.readthedocs.org/en/latest/details.html#how-good-is-assume

(I had a change where I had a strategy for `_FlakyAnnotation` itself, and then used that for populating the decorator, but thought that would diminish the utility of tests as examples)